cupy/core/core.pyx

```diff
@@ -137,13 +137,15 @@ cdef class ndarray:
             self.data = memory.alloc(self.size * itemsize)
         else:
             self.data = memptr
+        self._index_32_bits = (self.size * itemsize) <= (1 << 31)
```

Comment:
This condition is wrong in non-contiguous case. How is the following one?
```py
if self.data.mem.size <= (1 << 31):
    self._index_32_bits = True  # Fast path
else:
    bound = _memory_range._get_bound(array)
    self._index_32_bits = bound.second - bound.first < (1 << 31)
```

Reply:
We just discussed that this case only applies to views, but if a view associated array is  > 4GB the view will not be indexable with 32 bits indexing, (this will make it slower only).
So the current implementation is correct.

Reply:
This fix is needed when we provide the array with an external pointer and custom strides, there is the possibility of the strides to be bigger than 4GB